libs/core/src/service/document_service.rs

```diff
@@ -20,6 +23,23 @@ impl<Client: Requester> RequestContext<'_, '_, Client> {
 
         let doc = tree.read_document(self.config, &id, account)?;
 
+        let mut doc_events = self.tx.docs_events.get(&id).unwrap_or(&Vec::new()).clone();
+        doc_events.sort();
+        let latest_event = doc_events
+            .iter()
+            .filter(|e| matches!(e, DocEvents::Read(_)))
+            .last();
```

Comment:
I would use this [`last()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.last) instead, O(n) vs O(1)

Reply:
You don't have to filter for the last read, you can just check if the last is a read, same goes for write.